### PR TITLE
Functional options for kubeadm types

### PR DESCRIPTION
**Pull request by chuckha**
Signed-off-by: Chuck Ha <chuckh@vmware.com>

Heads up: the majority of this code is testing!

**What this PR does / why we need it**:
This PR improves the kubeadm configuration from a hard to follow non go-ish set of functions to something much more go-ish. This uses function options to modify and set values on the various kubeadm configurations.

I believe this will move us in a direction that can be reused in a kubeadm bootstrapper. This is essentially a library around kubeadm to generate kubeadm configurations in a composable way. 

I'll add some comments around the areas that probably need improving, but for now there is great code coverage, smaller functions and overall less code.

```release-note
NONE
```

### CODE REVIEW

File: pkg/cloud/aws/services/ec2/instances.go [hunk 1]
```diff
@@ -176,9 +177,23 @@ func (s *Service) createInstance(machine *actuators.MachineScope, bootstrapToken
...
 			s.scope.V(2).Info("Allowing a machine to join the control plane")
 
-			updatedJoinConfiguration := kubeadm.SetJoinNodeConfigurationOverrides(caCertHash, bootstrapToken, machine, &machine.MachineConfig.KubeadmConfiguration.Join)
-			updatedJoinConfiguration = kubeadm.SetControlPlaneJoinConfigurationOverrides(updatedJoinConfiguration)
-			joinConfigurationYAML, err := kubeadm.ConfigurationToYAML(updatedJoinConfiguration)
+			kubeadm.SetJoinConfigurationOptions(
+				&machine.MachineConfig.KubeadmConfiguration.Join,
+				kubeadm.WithBootstrapTokenDiscovery(
+					kubeadm.NewBootstrapTokenDiscoveryWithOptions(
+						kubeadm.WithAPIServerEndpoint(fmt.Sprintf("%s:%d", machine.Network().APIServerELB.DNSName, 6443)),
```
Comment (chuckha) [inline pkg/cloud/aws/services/ec2/instances.go]: Some of the constants should get pulled up, like 6443

File: pkg/cloud/aws/services/ec2/instances.go [hunk 2]
```diff
@@ -176,9 +177,23 @@ func (s *Service) createInstance(machine *actuators.MachineScope, bootstrapToken
...
+						kubeadm.WithToken(bootstrapToken),
+						kubeadm.WithCACertificateHash(caCertHash),
+					),
+				),
+				kubeadm.WithJoinNodeRegistrationOptions(
+					kubeadm.NewNodeRegistrationWithOptions(
+						kubeadm.WithTaints(machine.GetMachine().Spec.Taints),
+					),
+				),
+				kubeadm.WithLocalAPIEndpointAndPort("{{ ds.meta_data.local_ipv4 }}", 6443),
```
Comment (chuckha) [inline pkg/cloud/aws/services/ec2/instances.go]: and this constant should get pulled up and a comment should be added

**Comment by vincepri** [comment]
/assign @sethp-nr @rudoi 

for further review

File: pkg/cloud/aws/services/kubeadm/configs.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+			c.ControlPlaneEndpoint = endpoint
+		}
+	}
+}
+
+// WithAPIServerCertificateSANs will add the names provided to the kube-apiserver's list of certificate SANs.
+func WithAPIServerCertificateSANs(names ...string) ClusterConfigurationOption {
+	return func(c *kubeadmv1beta1.ClusterConfiguration) {
+		for _, name := range names {
+			c.APIServer.CertSANs = append(c.APIServer.CertSANs, name)
```
Comment (rudoi) [inline pkg/cloud/aws/services/kubeadm/configs.go]: I think `append(c.APIServer.CertSANs, names...)` could save some lines here? Not critical 🤷‍♂

Comment (chuckha) [reply pkg/cloud/aws/services/kubeadm/configs.go]: sure can! nice find
